// vtex/lib/lib/telemetry/TelemetryStore.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const configstore_1 = __importDefault(require("configstore"));
class TelemetryLocalStore {
    constructor(storeName) {
        this.storeName = storeName;
        this.store = new configstore_1.default(storeName);
    }
    getErrors() {
        return this.store.get('errors') || [];
    }
    getMetrics() {
        return this.store.get('metrics') || {};
    }
    getLastRemoteFlush() {
        var _a;
        return (_a = this.store.get('lastRemoteFlush')) !== null && _a !== void 0 ? _a : 0;
    }
    setErrors(errors) {
        this.store.set('errors', errors);
    }
    setMetrics(metrics) {
        this.store.set('metrics', metrics);
    }
    setLastRemoteFlush(date) {
        this.store.set('lastRemoteFlush', date);
    }
    clear() {
        this.store.clear();
    }
}
exports.TelemetryLocalStore = TelemetryLocalStore;
